source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+    }
+    else
+    {
+        AZLogWarn( ( "Request action called in wrong state %d\r\n",
+                     xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Implementation of parsing response action, this action is only allowed in azureiotprovisioningWF_STATE_RESPONSE
+ *
+ * */
+static void prvAzureIoTProvisioningClientParseResponse( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    az_result xCoreResult;
+    az_span xPayload = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponsePayload,
+                                       ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.xProvisioningLastResponsePayloadLength );
+    az_span xTopic = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponseTopic,
+                                     ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.usProvisioningLastResponseTopicLength );
+
+    /* Check the state.  */
+    if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_RESPONSE )
+    {
+        xCoreResult =
+            az_iot_provisioning_client_parse_received_topic_and_payload( &xAzureIoTProvisioningClientHandle->_internal.xProvisioningClientCore,
+                                                                         xTopic, xPayload, &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse );
+
+        if( az_result_failed( xCoreResult ) )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningFailed );
+            AZLogError( ( "IoTProvisioning client failed to parse packet, error status: %08x", xCoreResult ) );
+            return;
+        }
+
+        if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.operation_status == AZ_IOT_PROVISIONING_STATUS_ASSIGNED )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningSuccess );
+        }
+        else if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.retry_after_seconds == 0 )
+        {
+            AZLogError( ( "IoTProvisioning client failed with error %d, extended error status: %u\r\n",
+                          xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.error_code,
+                          xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.extended_error_code ) );
+            /* Server responded with error with no retry.  */
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningServerError );
+        }
+        else
+        {
+            xAzureIoTProvisioningClientHandle->_internal.ullProvisioningRequestTimeout =
```

Comment:
Is `ullProvisioningRequestTimeout` the right name for this?  It seems like the absolute time since you're doing a xGetTime() and adding the delta.  But this feels more like a `ulWhenToRetry` or something?

Reply:
ullProvisioningRetryAfter

---
In reply to: [602633333](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r602633333) [](ancestors = 602633333)